src/main/java/io/goobox/sync/common/overlay/OverlayHelper.java

```diff
@@ -174,13 +187,22 @@ public void refresh(Path path) {
                     .limit(syncDir.relativize(path).getNameCount())
                     .map(Path::toString)
                     .toArray(String[]::new);
-            fileIconControl.refreshIcons(pathAndParents);
+            refreshIcons(pathAndParents);
         }
     }
 
     private void refresh() {
         if (fileIconControl != null) {
-            fileIconControl.refreshIcons(new String[]{syncDir.toString()});
+            refreshIcons(new String[]{syncDir.toString()});
+        }
+    }
+
+    private void refreshIcons(String[] paths) {
+        try {
+            initializeResult.get();
```

Comment:
This will block the calling thread until the initialization finishes, won't it? This is something we want to avoid and the main reason for running the initialization in a separate thread.

It would be better if the method returns immediately without doing anything if the initialization is not complete yet.

Reply:
No no, you can do other stuffs while initializing overlay helper except for refreshing icons. 

Actually, in sync-sotrj, refreshIcons is called from only worker threads. So it won’t block something.

If this method returns immediately, how can the caller know the refreshing icons succeeds? The method can return `false` if the initialization hasn’t finished yet but in that case the caller of this method has to have a loop and retry it until it succeeds. Thus, waiting here until initialization finishes would be better.

Reply:
There is only one worker thread in the sync-storj app - `TaskExecutor`. It executes the tasks from a queue one after another. If a task wants to update an overlay icon and `refreshIcons()` blocks then the whole sync execution will block too. This is what I want to avoid.

The caller does not really care if the icon is updated or not. It just notifies that a file/folder should update its overlay icon. The caller does not even know if this file/folder is currently visible in the Windows Explorer. Calling `refreshIcons()` is such notification to the Liferay Nativity library. It's the responsibility of the latter to do the rest. If the library is not initialized yet then skip it and continue syncing. Once the library is initialized, I expect it would request the overlay icons for each visible file/folder in the Windows Explorer. If not, in the worst case, the user will have to hit F5.

Syncing is a the most important tasks and we should not block it due to decorations.

Reply:
Overlay icons are not the main GUI. The main GUI is the tray icon and the tray popup window, which should be able to properly display the current state even without this change. Overlay icons are helper GUI in the Windows Explorer and, as far as we have already seen, it's very likely that they won't work out-of-the-box for a major part of users. Thus, I really believe the top priority task for the app (synching) should not be blocked by displaying decorations in the file browser.

However, it's possible to have a solution that please both me and you. We can have a queue of `refreshIcons` requests. Calling the `refreshIcons()` method will only add an entry in the queue, but won't wait for the execution. There will be a dedicated thread to serve the queue - once the Liferay Nativity library is initialized the thread will start taking entries from the queue and process them. What do you think about this?

Reply:
I cannot agree that. For example, suppose both the system tray icon and popup window show all files are synchronized but one file doesn’t have any overlay icon. In that case, there are two possible situations:
- the file is synchronized but the overlay icon is missing,
- the file hasn’t been recognized by the app yet.  
It’s hard to distinguish those cases without reading the logs.

Anyway, using that queue is fine for me. Or just overwriting that method.

Reply:
Do you want me to make the change with introducing the queue? I have time right now.

Reply:
That’s okay. It’s not a big thing so I’ll implement it.